### ISSUE TITLE
Don't discourage to_f if objects could be strings

### ISSUE THREAD

**Issue by forthrin** [rubocop/rubocop#14628]
## Steps to reproduce the problem

Paste into terminal and press Ctrl+D

```bash
rubocop -s s
'3.14x2.71'.match(/(\S+)x(\S+)/)
puts $1.to_f / $2.to_f
puts $1.to_f / $2

```

## Expected behavior

No frisking.

## Actual behavior

**Advice**: `Style/FloatDivision: Prefer using .to_f on one side only.`
**Result**: `'Float#/': String can't be coerced into Float (TypeError)`

## RuboCop version

```
ruby      2025-08-01 30a20bc16
rubocop   2025-10-18 f7f653f50
System Version: macOS 26.0.1 (25A362)
Model Identifier: MacBookAir10,1
```

**Comment by Earlopain** [comment]
The cop is unsafe, this exact scenario is documented here https://docs.rubocop.org/rubocop/cops_style.html#safety-stylefloatdivision

**Comment by forthrin** [comment]
Interestingly, this cop is linked to division specifically and not superfluous `to_f` for other arithmetic. I'm actually having a hard time grasping why this cop graduated from PA at all, when prone to give code-breaking advice and cause less clarity to boot.

0. `a / b` # No signs of floats. Assume safe = **OK**
1. `a.to_f / b` # Float operations will cascade for all operations in scope = **OK**
2. `a.to_f / b.to_f` # Expressly states that both are floats, and might also do required conversion from string = **OK**

**Comment by koic** [comment]
For `Regexp.last_match(1)` and `$1`, it assumes that the value is a string matched by a regular expression, and allows conversion with `#to_f`. I've opened PR #14629 to resolve this issue.

**Comment by forthrin** [comment]
WFM! Though still not quite on board what problem it aspires to solve and thus why it exists in the first place. Why is it even enabled by default if it's unsafe? (Are others?)

```ruby
a = '3.14'
b = '2.71'
c = '1.61'
puts a.to_f / b.to_f / c.to_f

```

**Comment by koic** [comment]
RuboCop's static analysis can't really tell what kind of object a variable actually is. If it had a feature to track local variables, such checks would be possible, but that's not supported now. And once a variable becomes a method argument, tracking it becomes impossible.

Because of these limitations, RuboCop provides the concept of unsafe cops. Whether a cop should be enabled by default depends on how universal the rule is and how often false positives occur. So far, there haven't been many reports of such issues with `Style/FloatDivision` cop.

**Comment by forthrin** [comment]
Obviously there are limits to a linter's scope. Even a full interpreter cannot anticipate runtime anomalies with loose typing. Would appreciate if someone could list the rationales for its existence. If only removing fluff, the cons win this one.

**Comment by Earlopain** [comment]
There are many cops that are enabled and are also unsafe, that is nothing special. Probably the first introduction for it was here: https://github.com/rubocop/ruby-style-guide/issues/628.

I also never ran into issues with this cop. That said, I'm not a huge fan of it myself (or rather, the current default style). I don't mind `to_f` on both sides (possible to configure in such a way), even if both are integers. Coercion takes care of it but I always felt iffy about that. Explicit `to_f` twice just is more clear to me.

**Comment by bbatsov** [comment]
For the record - I plan to limit the number of enabled cops in the next major version of RuboCop (see #14005), but admittedly we've made pretty little progress towards it. 

The fact that unsafe cops are enabled by default is historical, and so if the fact that way too many cops (some of questionable value) are enabled by default. But such things can't be addressed in a backwards compatible fashion (as changing the list of default cops is backwards incompatible) and I think some notion of presets in the easiest way forward - it would provide simpler defaults, but also make it easy to stick to the old defaults. Technically speaking the presets can be introduced even now, we just can't change the "default" preset.
